kong/tools/utils.lua

```diff
@@ -270,13 +270,56 @@ do
     end
   end
 
+  local function compare_keys(a, b)
+    local ta = type(a)
+    if ta == type(b) then
+      return a < b
+    end
+    return ta == "number" -- numbers go first, then the rest of keys (usually strings)
+  end
+
+
+  -- Recursively URL escape and format key and value
+  -- Handles nested arrays and tables
+  local function recursive_encode_args(parent_key, value, raw, no_array_indexes, query)
+    local sub_keys = {}
+    for sk in pairs(value) do
+      sub_keys[#sub_keys + 1] = sk
+    end
+    sort(sub_keys, compare_keys)
+
+    local sub_value, next_sub_key
+    for _, sub_key in ipairs(sub_keys) do
+      sub_value = value[sub_key]
+
+      if no_array_indexes then
+        next_sub_key = parent_key
```

Comment:
Are we missing tests for `no_array_indexes` or did I miss them?

Reply:
You were completely right on this one, thanks.

There were no tests for this and it turns out that the implementation we had did not do it correctly (it should return `array[]` instead of `array[1]`, but it was returning `array` instead).

I added tests for `no_array_indexes`. I also made the admin api params parser able to cope with parameters with missing array indexes too, with tests.

Reply:
@kikito Great! It seems like a key-auth integration test is failing; likely due to a recent change though (it fails with form-urlencoded payloads)

Reply:
Yep, noticed yesterday but could not fix because I was busy preparing presentations. I have fixed key-auth on a separate commit. It is green now.